### PR TITLE
HAL API format

### PR DESCRIPTION
**Pull request by ahaith**
Return HAL format response on all API

### CODE REVIEW

**Comment by martinfilliau** [comment]
* [ ] The iCalendar representation is broken as the HAL serialisation has been introduced so there's a bit of work to adapt it. Tests should also be improved.
* [ ] Sort by `start_at` date #243
* [ ] Check the serialisation as (IMHO) the API should return (at least) a Javascript date, not just a formatted one

**Comment by ahaith** [comment]
When I get a search as a .ics and import the ics into Calendar, the end time is the same as the start time. It would be nice to include the actual end time.

**Comment by ahaith** [comment]
I get an error when trying to get the ics for an individual event:
http://talks.local:8000/api/events/banking-union.ics
"TypeError at /api/events/banking-union.ics
string indices must be integers, not str"

**Comment by ahaith** [comment]
There's a URL within the iCal feed but it doesn't include the stem - it's just the relative path from talks.ox.ac.uk

File: talks/core/renderers.py [hunk 1]
```diff
@@ -1,28 +1,46 @@
...
+        if 'title' in e:
+            event.add('summary', e['title'])
+        if 'description' in e:
+            event.add('description', e['description'])
+        if 'start' in e:
+            event.add('dtstart', dt_string_to_object(e['start']))
+        if 'end' in e:
+            event.add('dtend', dt_string_to_object(e['end']))
+        if 'url' in e:
+            event.add('url', e['url'])
```
Comment (ahaith) [inline talks/core/renderers.py]: This just returns the relative url which isn't much help outside of the site. Is it possible to generate the absolute url here from the request?

File: talks/core/renderers.py [hunk 2]
```diff
@@ -1,28 +1,46 @@
...
 
+        for e in data:
+            cal.add_component(self._event_to_ics(e))
         return cal.to_ical()
+
+    @staticmethod
+    def _event_to_ics(e):
+        event = Event()
+        if 'title' in e:
+            event.add('summary', e['title'])
```
Comment (ahaith) [inline talks/core/renderers.py]: Untitled talks end up as untitled calendar entries. We should perhaps add 'Oxford University talk' or something in this case.

File: talks/api/urls.py [hunk 3]
```diff
@@ -20,17 +21,19 @@ class TalksAPIRouter(routers.DefaultRouter):
     ]
 
 router = TalksAPIRouter()
-router.register(r'events', EventViewSet)
 router.register(r'series', EventGroupViewSet)
```
Comment (ahaith) [inline talks/api/urls.py]: It's currently still possible to download the calendar for a series with api/series/<slug>.ics. I think it's definitely desirable to have this functionality.
Currently, however the file downloaded doesn't seem to contain any events. I wonder if we should remove the Django REST viewset for these too and just directly specify our own url patterns and views as you've done for events.

File: talks/core/renderers.py [hunk 4]
```diff
@@ -1,28 +1,48 @@
...
+        if 'description' in e:
+            event.add('description', e['description'])
+        if 'start' in e:
+            event.add('dtstart', dt_string_to_object(e['start']))
+        if 'end' in e:
+            event.add('dtend', dt_string_to_object(e['end']))
+        if 'full_url' in e:
+            event.add('url', e['full_url'])
+            event.add('uid', e['full_url'])
+        # TODO add location field
```
Comment (ahaith) [inline talks/core/renderers.py]: Can we add the location? I guess just the api_location name would be enough.

**Pull request merged**